src/cascadia/Remoting/WindowManager.cpp

```diff
@@ -55,32 +33,178 @@ namespace winrt::Microsoft::Terminal::Remoting::implementation
         // monarch!
         CoRevokeClassObject(_registrationHostClass);
         _registrationHostClass = 0;
-        SignalClose();
-        _monarchWaitInterrupt.SetEvent();
+    }
 
-        // A thread is joinable once it's been started. Basically this just
-        // makes sure that the thread isn't just default-constructed.
-        if (_electionThread.joinable())
-        {
-            _electionThread.join();
-        }
+    void WindowManager::_createMonarch()
+    {
+        // Heads up! This only works because we're using
+        // "metadata-based-marshalling" for our WinRT types. That means the OS is
+        // using the .winmd file we generate to figure out the proxy/stub
+        // definitions for our types automatically. This only works in the following
+        // cases:
+        //
+        // * If we're running unpackaged: the .winmd must be a sibling of the .exe
+        // * If we're running packaged: the .winmd must be in the package root
+        _monarch = try_create_instance<Remoting::IMonarch>(Monarch_clsid,
+                                                           CLSCTX_LOCAL_SERVER);
+    }
+
+    // Check if we became the king, and if we are, wire up callbacks.
+    void WindowManager::_createCallbacks()
+    {
+        assert(_monarch);
+        // Here, we're the king!
+        //
+        // This is where you should do any additional setup that might need to be
+        // done when we become the king. This will be called both for the first
+        // window, and when the current monarch dies.
+
+        _monarch.WindowCreated({ get_weak(), &WindowManager::_WindowCreatedHandlers });
+        _monarch.WindowClosed({ get_weak(), &WindowManager::_WindowClosedHandlers });
+        _monarch.FindTargetWindowRequested({ this, &WindowManager::_raiseFindTargetWindowRequested });
+        _monarch.QuitAllRequested({ get_weak(), &WindowManager::_QuitAllRequestedHandlers });
+
+        _monarch.RequestNewWindow({ get_weak(), &WindowManager::_raiseRequestNewWindow });
+    }
+
+    void WindowManager::_registerAsMonarch()
+    {
+        winrt::check_hresult(CoRegisterClassObject(Monarch_clsid,

```

Comment:
wait, the PR says this exists to facilitate cross-window communication _but this is absolutely a cross-PROCESS registration_. Can you clarify?

Reply:
Sure.

We still need to do a _very minimal_ amount of x-proc calls. Namely, one right on startup, to see if another Terminal.exe was already running. If we find one, then we toss our commandline at it and bail. If we don't, then we need to CoRegister to prepare for subsequent launches to send commands to us.

Reply:
BTW optimally we should probably try to register first and then toss the commandline only if the registration failed. That avoids a race in between tossing and registering, when multiple WT instances start simultaneously.

Reply:
Added this to the description